mino/minows/session.go

```diff
@@ -2,85 +2,186 @@ package minows
+	"io"
+	"sync"
 )
 
-type envelope struct {
-	author mino.Address
-	msg    serde.Message
-	err    error
+var ErrWrongAddressType = xerrors.New("wrong address type")
+var ErrNotPlayer = xerrors.New("not player")
+
+// MaxUnreadAllowed Maximum number of unread messages allowed
+// in orchestrator's incoming message buffer before pausing relaying
+const MaxUnreadAllowed = 1e3
+
+type Forward struct {
+	Packet
+	Destination []byte
 }
 
-// session represents a stream session started by rpc.Stream()
-// between a minows instance and other players of the RPC.
-// A session ends for all participants when the initiator is done and cancels
-// the stream context.
-// - implements mino.Sender, mino.Receiver
-type session struct {
-	rpc  rpc
-	in   chan envelope
-	outs map[peer.ID]*json.Encoder
+type orchestrator struct {
+	logger zerolog.Logger
+
+	myAddr orchestratorAddr
+	rpc    rpc
+	// Connects to the participants
+	outs map[peer.ID]*gob.Encoder
+	in   chan Packet
 }
 
-// Send sends a message to all addresses concurrently.
-// Send is asynchronous and returns immediately an error channel that
-// closes when the message has either been sent or errored to each address.
-func (s session) Send(msg serde.Message, addrs ...mino.Address) <-chan error {
+func (o orchestrator) Send(msg serde.Message, addrs ...mino.Address) <-chan error {
 	send := func(addr mino.Address) error {
-		dest, ok := addr.(address)
-		if !ok {
-			return xerrors.Errorf("wrong address type: %T", addr)
+		var unwrapped address
+		switch a := addr.(type) {
+		case address:
+			unwrapped = a
+		case orchestratorAddr:
+			unwrapped = a.address
+		default:
+			return xerrors.Errorf("%v: %T", ErrWrongAddressType, addr)
 		}
-		out, ok := s.outs[dest.identity]
+
+		encoder, ok := o.outs[unwrapped.identity]
 		if !ok {
-			return xerrors.Errorf("address %v not a player", dest)
+			return xerrors.Errorf("%v: %v", ErrNotPlayer, addr)
+		}
+		src, err := o.myAddr.MarshalText()
+		if err != nil {
+			return xerrors.Errorf("could not marshal address: %v", err)
+		}
+		payload, err := msg.Serialize(o.rpc.context)
+		if err != nil {
+			return xerrors.Errorf("could not serialize message: %v", err)
 		}
-		return s.rpc.send(out, msg)
+
+		err = encoder.Encode(&Packet{Source: src, Payload: payload})
+		if err != nil {
+			return xerrors.Errorf("could not encode packet: %v", err)
+		}
+		return nil
 	}
 
-	result := make(chan envelope, len(addrs))
+	errs := doSend(addrs, send, msg, o.logger)
+	return errs
+}
+
+func (o orchestrator) Recv(ctx context.Context) (mino.Address, serde.Message, error) {
+	return doReceive(ctx, o.in, o.rpc.mino.GetAddressFactory(), o.rpc.factory,
+		o.rpc.context, o.logger)
+}
+
+type participant struct {
+	logger zerolog.Logger
+
+	myAddr address
+	rpc    rpc
+	// Connects to the orchestrator
+	out *gob.Encoder
+	in  chan Packet
+}
+
+func (p participant) Send(msg serde.Message, addrs ...mino.Address) <-chan error {
+	send := func(addr mino.Address) error {
```

Comment:
- Make this a method of `participant` - else it's capturing `p` here, which is confusing. If it's part of `participant`, it's clear that it will use the fields of `participant`
- Take `msg` as parameter, else the captured `msg` will be used both in this closure and be passed to `doSend`

Reply:
changed